File: ui/src/synergy/SynTraitValue.ts
import { startCase } from "lodash"
import { SynTraitHouse } from "../generated-src/SynTraitHouse"
import { SynTraitPlayer } from "../generated-src/SynTraitPlayer"
import { SynTraitValue } from "../generated-src/SynTraitValue"
import { Utils } from "../config/Utils"

export type SynTraitRatingValues = -4 | -3 | -2 | -1 | 1 | 2 | 3 | 4

export const synTraitValueToString = (value: SynTraitValue) => `${value.trait} – ${value.rating} – ${value.house}`

export const synTraitName = (traitValue: SynTraitValue) => {
    const {
        trait,
        player,
        cardTypes,
        fromZones,
        powersString,
        cardTraits,
        cardName,
        house,
        synergyGroup,
        synergyGroupMax,
        notCardTraits,
        primaryGroup
    } = traitValue
    let name: string = trait

    if (cardName != null) {
        name = cardName
    }
    let nameEnhancer = ""
    if (powersString.length > 0) {
        nameEnhancer += ` ${powersString} Power`
    }
    if (player && player != SynTraitPlayer.ANY) {
        nameEnhancer += ` ${startCase(player.toLowerCase())}`
    }
    if (cardTraits != null && cardTraits.length > 0) {
        nameEnhancer += ` ${notCardTraits ? "Non-" : ""}${Utils.generateTraitsTypeList(cardTraits)}`
    }
    if (cardTypes && cardTypes.length > 0) {
        if (cardTypes.length === 1) {
            nameEnhancer += ` ${cardTypes[0]}s`
        } else {
            nameEnhancer += ` ${Utils.generateTraitsTypeList(cardTypes, "s")}`
        }
    }
    if (name.includes("_R_")) {
        name = startCase(name).replace(" R ", nameEnhancer + " ")
    } else {
        name = startCase(name) + " " + nameEnhancer
    }
    if (house === SynTraitHouse.continuous) {
        name = "Omni: " + name
    }
    if (fromZones && fromZones.length > 0) {
        if (fromZones.length === 1) {
            name += ` from ${Utils.camelCaseToTitleCase(fromZones[0])}`
        } else {
            const zonesUpper = fromZones.map(zone => Utils.camelCaseToTitleCase(zone))
            name += ` from ${Utils.generateTraitsTypeList(zonesUpper)}`
        }
    }
    if (synergyGroup != null) {
        name += ` – ${synergyGroup}${primaryGroup ? "⭑" : ""}`
        if (synergyGroupMax != null) {
            name += ` ${synergyGroupMax > 0 ? "<" : ">"}${synergyGroupMax}%`
        }
    }
    return name.trim()
}

File: src/main/kotlin/coraythan/keyswap/synergy/SynTraitValue.kt
package coraythan.keyswap.synergy

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import coraythan.keyswap.cards.CardType
import coraythan.keyswap.cards.extrainfo.ExtraCardInfo
import coraythan.keyswap.generatets.GenerateTs
import coraythan.keyswap.generatets.TsIgnore
import coraythan.keyswap.generatets.TsOptional
import coraythan.keyswap.startCase
import io.hypersistence.utils.hibernate.type.array.ListArrayType
import jakarta.persistence.*
import org.hibernate.annotations.Parameter
import org.hibernate.annotations.Type
import java.util.*

@GenerateTs
@Entity
data class SynTraitValue(
    @Enumerated(EnumType.STRING)
    val trait: SynergyTrait,
    val rating: Int = 3,
    @Enumerated(EnumType.STRING)
    val house: SynTraitHouse = SynTraitHouse.anyHouse,

    @Enumerated(EnumType.STRING)
    val player: SynTraitPlayer = SynTraitPlayer.ANY,

    @Type(
        value = ListArrayType::class,
        parameters = [Parameter(
            value = "card_type",
            name = ListArrayType.SQL_ARRAY_TYPE,
        )]
    )
    @Column(columnDefinition = "card_type[]")
    val cardTypes: List<CardType>? = listOf(),

    @Type(ListArrayType::class)
    @Column(columnDefinition = "text[]")
    val cardTraits: List<String>? = listOf(),

    @TsIgnore
    var cardTraitsString: String = "",
    @TsIgnore
    var cardTypesString: String = "",

    /**
     * Creature power expression, one of:
     *
     * ['odd','even','2-5','2 or less','3+','3,5,7']
     */
    var powersString: String = "",
    val notCardTraits: Boolean = false,

    @TsIgnore
    @JsonIgnoreProperties("traits")
    @ManyToOne
    val traitInfo: ExtraCardInfo? = null,

    @TsIgnore
    @JsonIgnoreProperties("synergies")
    @ManyToOne
    val synergyInfo: ExtraCardInfo? = null,

    val cardName: String? = null,

    val synergyGroup: String? = null,
    val synergyGroupMax: Int? = null,
    val primaryGroup: Boolean = false,

    @TsOptional
    @Id
    val id: UUID = UUID.randomUUID()
) : Comparable<SynTraitValue> {
    override fun compareTo(other: SynTraitValue): Int {
        return other.rating - this.rating
    }

    fun strength() = when (rating) {
        1 -> TraitStrength.EXTRA_WEAK
        2 -> TraitStrength.WEAK
        3 -> TraitStrength.NORMAL
        4 -> TraitStrength.STRONG
        else -> TraitStrength.NORMAL
    }

    override fun toString(): String {
        return "SynTraitValue(trait=$trait, rating=$rating, type=$house, player=$player, cardTypes=$cardTypesString traitInfoId=${traitInfo?.id}, synergyInfoId=${synergyInfo?.id}, id=$id)"
    }

    fun powerMatch(power: Int, cardType: CardType? = CardType.Creature): Boolean {
        return when {
            powersString.isBlank() -> {
                true
            }

            cardType != CardType.Creature && cardType != CardType.TokenCreature -> {
                false
            }

            powersString == "even" -> {
                power % 2 == 0
            }

            powersString == "odd" -> {
                power % 2 != 0
            }

            powersString.contains(" or less") -> {
                power <= powersString.replace(" or less", "").toInt()
            }

            powersString.contains("+") -> {
                power >= powersString.replace("+", "").toInt()
            }

            powersString.contains("-") -> {
                val nums = powersString.split("-")
                power >= nums[0].toInt() && power <= nums[1].toInt()
            }

            powersString.contains(",") -> {
                val nums = powersString.split(",")
                nums.any { power == it.toInt() }
            }

            powersString.toIntOrNull() != null -> {
                powersString.toInt() == power
            }

            else -> {
                throw IllegalStateException("Invalid power string: $powersString")
            }
        }
    }

    fun validate() {
        powerMatch(1)
    }

    fun print(): String {
        var name = if (trait == SynergyTrait.any) "" else trait.toString()
        if (cardName != null) {
            name = cardName
        }
        var nameEnhancer = ""
        if (powersString.isNotBlank()) {
            nameEnhancer += " $powersString Power"
        }
        if (player != SynTraitPlayer.ANY) {
            nameEnhancer += " ${player.toString().lowercase().replaceFirstChar { cap -> cap.uppercase() }}"
        }
        if (!cardTraits.isNullOrEmpty()) {
            nameEnhancer += " ${if (notCardTraits) "Non-" else ""}${
                cardTraits.joinToString(", ") {
                    it.lowercase().replaceFirstChar { cap -> cap.uppercase() }
                }
            }"
        }
        if (!cardTypes.isNullOrEmpty()) {
            nameEnhancer += if (cardTypes.size == 1) {
                " ${cardTypes[0]}s"
            } else {
                " ${cardTypes.joinToString(" ") { "${it}s" }}"
            }
        }
        name = if (name.contains("_R_")) {
            name.startCase().replace("_ R_", nameEnhancer)
        } else {
            "${name.startCase()} $nameEnhancer"
        }
        name = when (house) {
            SynTraitHouse.anyHouse -> ""
            SynTraitHouse.continuous -> "Omni: "
            SynTraitHouse.outOfHouse -> "Out of House: "
            SynTraitHouse.house -> "In House: "
        } + name

        name += when (rating) {
            1 -> " +"
            2 -> " ++"
            3 -> " +++"
            4 -> " ++++"
            -1 -> " -"
            -2 -> " --"
            -3 -> " ---"
            -4 -> " ----"
            else -> ""
        }

        return name
    }

}

fun Collection<SynTraitValue>.containsTrait(trait: SynergyTrait, player: SynTraitPlayer? = null) =
    this.find { valueTrait ->
        valueTrait.trait == trait && (player == null || player == valueTrait.player)
    } != null
